velox/exec/tests/utils/TestIndexStorageConnector.cpp

```diff
@@ -0,0 +1,258 @@
+      pool_(connectorQueryCtx_->memoryPool()->shared_from_this()),
+      executor_(executor) {
+  VELOX_CHECK_NOT_NULL(executor_);
+  VELOX_CHECK(inputType_->equivalent(*keyType_));
+  VELOX_CHECK_LE(outputType_->size(), valueType_->size());
+  // As for now, we only support to read data columns.
+  for (int i = 0; i < outputType_->size(); ++i) {
+    VELOX_CHECK(outputType_->childAt(i)->equivalent(
+        *valueType_->childAt(valueType_->getChildIdx(outputType_->nameOf(i)))));
+  }
+  // As for now we require the all the index columns to be specified in lookup
+  // input.
+  for (int i = 0; i < keyType_->size(); ++i) {
+    VELOX_CHECK(keyType_->childAt(i)->equivalent(*inputType_->childAt(i)));
+  }
+  initOutputProjections();
+}
+
+std::unique_ptr<connector::IndexSource::LookupResultIterator>
+TestIndexSource::lookup(const LookupRequest& request) {
+  const auto numIputRows = request.input->size();
+  auto& hashTable = tableHandle_->indexTable()->table;
+  auto lookup = std::make_unique<HashLookup>(hashTable->hashers());
+  SelectivityVector activeRows(numIputRows);
+  VELOX_CHECK(activeRows.isAllSelected());
+  hashTable->prepareForJoinProbe(
+      *lookup, request.input, activeRows, /*decodeAndRemoveNulls=*/true);
+  lookup->hits.resize(numIputRows);
+  std::fill(lookup->hits.data(), lookup->hits.data() + numIputRows, nullptr);
+  if (!lookup->rows.empty()) {
+    hashTable->joinProbe(*lookup);
+  }
+  // Update lookup rows to include all input rows as it might be used by left
+  // join.
+  auto& rows = lookup->rows;
+  rows.resize(request.input->size());
+  std::iota(rows.begin(), rows.end(), 0);
+  return std::make_unique<ResultIterator>(
+      this->shared_from_this(),
+      request,
+      std::move(lookup),
+      tableHandle_->asyncLookup() ? executor_ : nullptr);
+}
+
+void TestIndexSource::initOutputProjections() {
+  VELOX_CHECK(lookupOutputProjections_.empty());
+  lookupOutputProjections_.reserve(outputType_->size());
+  for (auto outputChannel = 0; outputChannel < outputType_->size();
+       ++outputChannel) {
+    auto tableValueChannel =
+        valueType_->getChildIdx(outputType_->nameOf(outputChannel));
+    // The hash table layout is: index columns, value columns.
+    lookupOutputProjections_.emplace_back(
+        keyType_->size() + tableValueChannel, outputChannel);
+  }
+  VELOX_CHECK_EQ(lookupOutputProjections_.size(), outputType_->size());
+}
+
+TestIndexSource::ResultIterator::ResultIterator(
+    std::shared_ptr<TestIndexSource> source,
+    const LookupRequest& request,
+    std::unique_ptr<HashLookup> lookupResult,
+    folly::Executor* executor)
+    : source_(std::move(source)),
+      indexTable_(source_->tableHandle_->indexTable()),
+      request_(request),
+      lookupResult_(std::move(lookupResult)),
+      executor_(executor) {
+  // Initialize the lookup result iterator.
+  std::vector<vector_size_t> varSizeListColumns;
+  std::iota(varSizeListColumns.begin(), varSizeListColumns.end(), 0);
```

Comment:
What does 'varSizeListColumns' stand for? It looks like this is an empty vector, no? What does std::iota do?

Reply:
It is used by hash probe to estimate the output batch size in production. I just add a fake estimated size here for later improvement which is not relevant for current implementation.

Reply:
@xiaoxmeng Are you going to remove this variable and std::iota call?

Reply:
yes, it is used to control the number of output rows from the list join which control based on the max rows and max bytes. Currently, we don't support bytes based output control in lookup so it is not required and just pass one byte for simplicity. Once we support bytes based output size control, we will revisit here. Thanks!